src/Compilers/CSharp/Portable/Lowering/LocalRewriter/DelegateCacheRewriter.cs

```diff
@@ -0,0 +1,166 @@
+
+        var oldSyntax = _factory.Syntax;
+        _factory.Syntax = syntax;
+
+        var cacheContainer = GetOrAddCacheContainer(localFunctionOrdinal, delegateType, targetMethod);
+        var cacheField = cacheContainer.GetOrAddCacheField(_factory, delegateType, targetMethod);
+
+        var boundCacheField = _factory.Field(null, cacheField);
+        var boundDelegateCreation = new BoundDelegateCreationExpression(syntax, receiver, targetMethod, isExtensionMethod: false, type: delegateType);
+
+        var rewrittenNode = _factory.Coalesce(boundCacheField, _factory.AssignmentExpression(boundCacheField, boundDelegateCreation));
+
+        _factory.Syntax = oldSyntax;
+
+        return rewrittenNode;
+    }
+
+    private DelegateCacheContainer GetOrAddCacheContainer(int localFunctionOrdinal, TypeSymbol delegateType, MethodSymbol targetMethod)
+    {
+        Debug.Assert(_factory.ModuleBuilderOpt is { });
+        Debug.Assert(_factory.CurrentFunction is { });
+
+        var typeCompilationState = _factory.CompilationState;
+        var generation = _factory.ModuleBuilderOpt.CurrentGenerationOrdinal;
+
+        DelegateCacheContainer? container;
+
+        if (CanUseConcreteCacheContainer(delegateType, targetMethod))
+        {
+            container = typeCompilationState.ConcreteDelegateCacheContainer;
+
+            if (container is { })
+            {
+                return container;
+            }
+
+            container = new(typeCompilationState.Type, generation);
+            typeCompilationState.ConcreteDelegateCacheContainer = container;
+        }
+        else
+        {
+            var containersStack = _genericCacheContainers ??= new();
+
+            while (containersStack.Count > 0 && containersStack.Peek().LocalFunctionOrdinal > localFunctionOrdinal)
+            {
+                containersStack.Pop();
+            }
+
+            if (containersStack.Count > 0 && containersStack.Peek().LocalFunctionOrdinal == localFunctionOrdinal)
+            {
+                return containersStack.Peek().CacheContainer;
+            }
+
+            container = new(_factory.CurrentFunction, _topLevelMethodOrdinal, localFunctionOrdinal, generation);
+            containersStack.Push((localFunctionOrdinal, container));
+        }
+
+        _factory.AddNestedType(container);
+
+        return container;
+    }
+
+    private bool CanUseConcreteCacheContainer(TypeSymbol delegateType, MethodSymbol targetMethod)
+    {
+        // Possible places for type parameters that can act as type arguments to construct the delegateType or targetMethod:
+        //   1. containing types
+        //   2. top level method
+        //   3. local functions
+        // Since our containers are created within the same enclosing types, we can ignore type parameters from them.
+
+        Debug.Assert(_factory.CurrentFunction is { });
+
+        // Obviously,
+        if (_topLevelMethod.Arity == 0 && (object)_factory.CurrentFunction == _topLevelMethod)
```

Comment:
>if (_topLevelMethod.Arity == 0 && (object)_factory.CurrentFunction == _topLevelMethod) [](http://example.com/codeflow?start=8&length=86)

I am not sure if it is worth checking this explicitly. It looks like the implementation would do the right thing without this check

Reply:
It can save some potential HashSet allocations?

Reply:
> It can save some potential HashSet allocations?

This feels like a premature optimization. Getting a hashset from a pull should be cheap. Also, we could postpone actual allocation until we have something to put into the hashset.

Reply:
Interesting! I'll remove this check then.